chirho/robust/handlers/estimators.py

```diff
@@ -33,9 +46,84 @@ def one_step_correction(
     """
     influence_kwargs_one_step = influence_kwargs.copy()
     influence_kwargs_one_step["pointwise_influence"] = False
-    eif_fn = influence_fn(model, guide, functional, **influence_kwargs_one_step)
+    eif_fn = influence_fn_estimator(
+        model, guide, functional, **influence_kwargs_one_step
+    )
 
     def _one_step(test_data: Point[T], *args, **kwargs) -> S:
         return eif_fn(test_data, *args, **kwargs)
 
     return _one_step
+
+
+def one_step_corrected_estimator(
+    model: Callable[P, Any],
+    guide: Callable[P, Any],
+    functional: Optional[Functional[P, S]] = None,
+    influence_fn_estimator: Callable[
+        [Callable[P, Any], Callable[P, Any], Optional[Functional[P, S]]],
+        Callable[Concatenate[Point[T], P], S],
+    ] = influence_fn,
+    **influence_kwargs,
+) -> Callable[Concatenate[Point[T], P], S]:
+    """
+    Returns a function that computes the one-step corrected estimator for the functional at a
+    specified set of test points as discussed in [1].
+
+    :param model: Python callable containing Pyro primitives.
+    :type model: Callable[P, Any]
+    :param guide: Python callable containing Pyro primitives.
+        Must only contain continuous latent variables.
+    :type guide: Callable[P, Any]
+    :param functional: model summary of interest, which is a function of the
+        model and guide. If ``None``, defaults to :class:`PredictiveFunctional`.
+    :type functional: Optional[Functional[P, S]], optional
+    :param influence_fn_estimator: function to approximate the efficient influence
+        function. Defaults to :func:`influence_fn`.
+    :type influence_fn_estimator: Callable[
+        [Callable[P, Any], Callable[P, Any], Optional[Functional[P, S]]],
+        Callable[Concatenate[Point[T], P], S],
+    ]
+    :return: function to compute the one-step corrected estimator
+    :rtype: S
+    """
+    if functional is None:
+        assert isinstance(model, torch.nn.Module)
+        assert isinstance(guide, torch.nn.Module)
+        plug_in_estimator = PredictiveFunctional(model, guide)
+    else:
+        plug_in_estimator = functional(model, guide)
+    correction_estimator = one_step_correction(
+        model, guide, functional, influence_fn_estimator, **influence_kwargs
+    )
+
+    def _one_step_corrected_estimator(test_data: Point[T], *args, **kwargs) -> S:
+        plug_in_estimate = plug_in_estimator(*args, **kwargs)
+        correction = correction_estimator(test_data, *args, **kwargs)
+        return apply_correction(plug_in_estimate, correction)
```

Comment:
Instead of defining a bespoke `apply_correction` and having to fight the type checker and handle a bunch of edge cases yourself, you might try using PyTorch's beta [`PyTree`](https://github.com/pytorch/pytorch/blob/main/torch/utils/_pytree.py) implementation, which is designed for precisely this problem:
```py
flat_plug_in_estimate, treespec = torch.utils._pytree.tree_flatten(plug_in_estimate)
flat_correction, _ = torch.utils._pytree.tree_flatten(correction)
return torch.utils._pytree.tree_unflatten([a + b for a, b in zip(flat_plug_in_estimate, flat_correction)], treespec)
```
The API is still changing a lot, but I think `tree_flatten` and `tree_unflatten` are stable.

Reply:
Cool, I'll give that a shot.